src/metrics/metric_manager.py

```diff
@@ -0,0 +1,177 @@
+
+
+class MetricParams:
+    """
+    Class for contain metric parameters.
+    """
+    def __init__(self, class_name: str, target_fields: dict, phases: List[str] = None, \
+                 name: str = None, metric_params: dict = {}):
+        """
+        Args
+            class_name: Metric class name with would be created.
+            target_fields: Dictionary for mapping Task output with Metric forward keys.
+            phases: Metric run phases.
+            name: Metric name in logging output.
+            metric_params: Metric class initialize parameters.
+        """
+        self.class_name = class_name
+        self.target_fields = target_fields
+        self.phases = phases if phases is not None else ['train', 'valid', 'test']
+        self.name = name
+        self.metric_params = metric_params
+
+
+class MetricWithUtils(nn.Module):
+    """
+    Union class for metric and metric utils parameters
+    """
+    def __init__(self, metric: Metric, target_fields: Dict[str, str], name: str = None):
+        """
+        Args:
+            metric: Metric written with TorchMetrics.
+            target_fields: Dictionary for mapping Metric forward input keys with Task output dictionary keys.
+            name: The metric name.
+        """
+        super().__init__()
+        self.__metric = metric
+        self.__target_fields = target_fields
+        self.__name = name
+
+    @property
+    def name(self):
+        return self.__name
+
+    @property
+    def target_fields(self):
+        return self.__target_fields
+
+    def forward(self, *args, **kwargs):
+        return self.__metric(*args, **kwargs)
+
+    def compute(self):
+        return self.__metric.compute()
+
+
+class MetricManager(nn.Module):
+    """Manages all metrics for the model."""
+
+    # model use phases
+    phases = ['train', 'valid', 'test']
+
+    def __init__(self, params: List[MetricParams]):
+        """
+        Args:
+            params: Metric parameters.
+        """
+        super().__init__()
+        phase2metrics = {phase: {} for phase in self.phases}
+        for phase in self.phases:
+            phase2metrics[phase] = self.__get_phase_metrics(params, phase)
+
+        self.phase2metrics = phase2metrics
+
+    def __get_phase_metrics(self, params: List[MetricParams], phase: str) -> nn.ModuleList:
+        """
+        Generate metric list for current phase.
+
+        Args:
+            params: All metric params from config file.
+            phase: Current phase name.
+
+        Return:
+            Metric list as nn.ModuleList for current phase. 
+        """
+        # create added_metric_names set
+        added_metric_names = set()
+        metrics = []
+        for metric_params in params:
+            if phase not in metric_params.phases:
+                continue
+            metric = METRICS.get(metric_params.class_name)(**metric_params.metric_params)
+            target_fields = metric_params.target_fields
```

Comment:
May be should remove target_fields and add target and prediction

Reply:
I think exist metric with more than one input argument. Moreover if we do that it force me to rewrite all torchmetrics metrics.